panda/src/gobj/texture.cxx

```diff
@@ -10349,20 +10365,27 @@ do_fillin_rawdata(CData *cdata, DatagramIterator &scan, BamReader *manager) {
       cdata->_ram_images[n]._page_size = scan.get_uint32();
     }
 
-    // fill the cdata->_image buffer with image data
-    size_t u_size = scan.get_uint32();
+    // fill the cdata->_image buffer with image data.
+    // Decompress if neccesary
+    size_t read_size = scan.get_uint32();
 
     // Protect against large allocation.
-    if (u_size > scan.get_remaining_size()) {
+    if (read_size > scan.get_remaining_size()) {
       gobj_cat.error()
         << "RAM image " << n << " extends past end of datagram, is texture corrupt?\n";
       return;
     }
 
-    PTA_uchar image = PTA_uchar::empty_array(u_size, get_class_type());
-    scan.extract_bytes(image.p(), u_size);
+    PTA_uchar image_data = PTA_uchar::empty_array(read_size, get_class_type());
+    scan.extract_bytes(image_data.p(), read_size);
+
+    if(manager->get_texture_compression_level() != BamEnums::BTC_off) {
```

Comment:
We should explicitly check for BTC_on here to make it easier to add new formats in the future. Also, we should throw a meaningful error if the compression format/level requested is not supported.

Reply:
Do we have to throw errors in both `texture.cxx` and `bamReader.cxx`? Throwing error in `bamReader::init()` should be enough.

Reply:
I would say yes. This keeps the code better de-coupled. If something changes in `bamReader::init()` that allows us to get to texture.cxx with bad data, then we'll still want to error. Or, if someone finds some way to call into the texture.cxx code without first going through `bamReady::init()`.